File: java/src/main/java/io/fivexer/sdk/model/NotificationSequenceStep.java
package io.fivexer.sdk.model;

import io.fivexer.sdk.internal.Json;

/**
 * One step of a notification sequence: fire {@code eventType} at {@code offsetMs} after
 * {@code trigger}. Used both when creating a sequence and when reading one back.
 */
public class NotificationSequenceStep {
    private String trigger;   // matched | expiring | expired | accepted | rejected | completed
    private Long offsetMs;
    private String eventType;

    /** Response construction (Gson). */
    public NotificationSequenceStep() {}

    public NotificationSequenceStep(String trigger, String eventType) {
        this.trigger = trigger;
        this.eventType = eventType;
    }

    public String getTrigger() { return trigger; }
    public Long getOffsetMs() { return offsetMs; }
    public String getEventType() { return eventType; }

    public NotificationSequenceStep offsetMs(long offsetMs) { this.offsetMs = offsetMs; return this; }

    /** Serialise for the wire, omitting the offset when it was never set. */
    public String toJson() {
        return Json.write(this);
    }
}

File: php/src/Model/NotificationSequenceStep.php
<?php

declare(strict_types=1);

namespace Fivexer\SDK\Model;

use Fivexer\SDK\Internal\Json;

/**
 * One step of a notification sequence: fire `eventType` at `offsetMs` after `trigger`.
 *
 * Serves as both an input model (build one, hand it to a sequence) and a response model
 * ({@see self::fromArray()}), since the shape is identical in both directions.
 */
final class NotificationSequenceStep
{
    public function __construct(
        private readonly string $trigger,
        private readonly string $eventType,
        private ?int $offsetMs = null,
    ) {
    }

    public function offsetMs(int $offsetMs): self
    {
        $this->offsetMs = $offsetMs;
        return $this;
    }

    public function getTrigger(): string
    {
        return $this->trigger;
    }

    public function getEventType(): string
    {
        return $this->eventType;
    }

    public function getOffsetMs(): ?int
    {
        return $this->offsetMs;
    }

    /** @return array<string, mixed> */
    public function toArray(): array
    {
        return [
            'trigger' => $this->trigger,
            'eventType' => $this->eventType,
        ] + Json::compact(['offsetMs' => $this->offsetMs]);
    }

    /**
     * @param array<string, mixed> $data
     * @internal
     */
    public static function fromArray(array $data): self
    {
        return new self(
            trigger: (string) ($data['trigger'] ?? ''),
            eventType: (string) ($data['eventType'] ?? ''),
            offsetMs: isset($data['offsetMs']) ? (int) $data['offsetMs'] : null,
        );
    }
}
